bot/utils/voice_recognition.py

```diff
@@ -0,0 +1,101 @@
+import os
+from google.cloud import speech
+import requests
+import logging
+import json
+from tempfile import NamedTemporaryFile
+import subprocess
+
+TOKEN = os.environ["TOKEN"]
+
+LANG = "ru-RU"
+SAMPLE_RATE_HERTZ = 48000
+ENCODING_TYPE = "LINEAR16"
+CHANNEL_COUNT = 1
+ENABLE_AUTOMATIC_PUNCTUATION = True
+
+logger = logging.getLogger(__name__)
+
+speechClient = speech.SpeechClient()
+
+
+def _get_audio_file_data(file_id):
+    """
+    gets audio file 
+    """
+    url = f"https://api.telegram.org/bot{TOKEN}/getFile?file_id={file_id}"
+    response = requests.get(url)
+    jsonResponse = response.json()
+    logger.info(json.dumps(jsonResponse, indent=4))
+    res = jsonResponse["result"]
+    return res
+
+
+def _get_binary_content(file):
+    file_name = file["file_path"]
+    url = f"https://api.telegram.org/file/bot{TOKEN}/{file_name}"
+    response = requests.get(url, stream=True)
+
+    suffix = ".oga"
+    tmp_voice = NamedTemporaryFile(delete=False, suffix=suffix)
+    tmp_voice.write(response.content)
+    tmp_voice_filename_ogg = tmp_voice.name
+    logger.info(f"Name of tempfile {tmp_voice_filename_ogg}")
+    tmp_voice_filename_wav = tmp_voice_filename_ogg[:-len(suffix)] + ".wav"
+    logger.info(f"Name of new tempfile {tmp_voice_filename_wav}")
+
+    subprocess.call(['ffmpeg', '-i', tmp_voice_filename_ogg, tmp_voice_filename_wav])
+
+    converted_wav = open(tmp_voice_filename_wav, "rb")
+    converted_wav_bin = converted_wav.read()
+
+    tmp_voice.close()
+    converted_wav.close()
+
+    os.remove(tmp_voice_filename_ogg)
+    os.remove(tmp_voice_filename_wav)
+
+    return converted_wav_bin
+
+
+def _send_binary_to_google_speech(content):
+    """
+    sends binary data of voice to google speech
+    """
+    config = {
+        "language_code": LANG,
+        "sample_rate_hertz": int(SAMPLE_RATE_HERTZ),
+        "encoding": ENCODING_TYPE,
+        "audio_channel_count": int(CHANNEL_COUNT),
+        "enable_automatic_punctuation": bool(ENABLE_AUTOMATIC_PUNCTUATION),
+        "model": "default",
+    }
+    audio = {"content": content}
+    speech_response = speechClient.recognize(config=config, audio=audio)
+    return speech_response
+
+
+def _check_google_speech_response(response):
+    result = response.results[-1]
+    logger.info(result)
+    any_alternatives = result.alternatives or []
+    if len(any_alternatives) > 0:
+        transcription = any_alternatives[0].transcript
+        return transcription
+    return None
+
+
+def get_text_from_speech(file_id):
+    """
+    gets text from voice
+    """
+    try:
+        file = _get_audio_file_data(file_id)
+        content = _get_binary_content(file)
+        response = _send_binary_to_google_speech(content)
+        result = _check_google_speech_response(response)
+        logger.info(f"{result}")
+        return result
+    except Exception as ex: 
```

Comment:
Too common Exception type. Is it possible to define more particular ones?

Reply:
I'm not a full python developer😄  Probably you could give me an advice, which approach to use here to resolve this discussion?

Reply:
Usually, you should define which exceptions are expectable, and try to catch them. An unexpectable exception must be raised to be caught outside a current abstraction layer.

Reply:
Done, please check.

Reply:
🔥 🔥 🔥 🔥 🔥 🔥